### ISSUE TITLE
NewDecoder Implementation gone. Tests broken.

### ISSUE THREAD

**Issue by dsoprea** [go-yaml/yaml#359]
It's not clear what the state of the io.Reader-based decoder is. The documentation (https://godoc.org/gopkg.in/yaml.v2#Decoder) says there is a `NewDecoder` and the tests (https://github.com/go-yaml/yaml/blob/v2.2.1/decode_test.go#L766) say there is a `NewDecoder`, but the implementation only has a `newDecoder` (https://github.com/go-yaml/yaml/blob/v2.2.1/decode.go#L243). I can't help but notice that there are no CI badges in Github so this implies that the tests are not being regularly run. When I attempt to run them, they appear to be broken:

```
$ go test -v gopkg.in/yaml.v2
# gopkg.in/yaml.v2
decode_test.go:5:2: cannot find package "gopkg.in/check.v1" in any of:
	/usr/local/go/src/gopkg.in/check.v1 (from $GOROOT)
	/home/dustin/development/gopath/src/gopkg.in/check.v1 (from $GOPATH)
FAIL	gopkg.in/yaml.v2 [setup failed]
```

**It seems like that, at this moment, there is no way to avoid reading the entire byte stream prior to being able to parse it.**

**Comment by niemeyer** [comment]
There's CI here. The badge is just a picture shown in the README. Feel free to submit a change adding it.

The error you show is just the lack of a package in your local system. Try installing it and let us know what happens.

**Comment by dsoprea** [comment]
You skipped over the first, bigger issue. Where is `NewDecoder`? Why doesn't the project support marshaling from an `io.Reader` any more?

For the second, you assumed that I didn't do a go-get. Did you even try?

```
$ rm -fr /tmp/gopath/
$ GOPATH=/tmp/gopath go get gopkg.in/yaml.v2
$ GOPATH=/tmp/gopath go test gopkg.in/yaml.v2
# gopkg.in/yaml.v2
/tmp/gopath/src/gopkg.in/yaml.v2/decode_test.go:11:2: cannot find package "gopkg.in/check.v1" in any of:
    /usr/local/go/src/gopkg.in/check.v1 (from $GOROOT)
    /tmp/gopath/src/gopkg.in/check.v1 (from $GOPATH)
FAIL    gopkg.in/yaml.v2 [setup failed]
```

Yes. I'm aware the badge is just an image. Traditionally, the badge is used if there's any CI. It's not a guarantee, though. That's why I said the *lack* of a badge *and* broken unit-tests *implied* there was no CI. Does that make sense? 

I would also request that you re-open the issue while we're discussing it so that the community can be fully aware/involved.

**Comment by rogpeppe** [comment]
Try `go get -t gopkg.in/yaml.v2`.

**Comment by niemeyer** [comment]
You speak of NewDecoder but post a completely unrelated error. That error is what I explained. What are the steps to reproduce the error about NewDecoder and what is the error?

**Comment by niemeyer** [comment]
Also unit tests are not broken, and we have CI.

**Comment by dsoprea** [comment]
Gustavo, in the first message that I wrote, I included links and mentioned why I said what I said. Please read it carefully. I mentioned the tests in addition to that. Those were two different things I mentioned in the same place.

I recognize that I might be wrong, but please show me how. It's a community and we're all just trying to help. Be patient with me, please.

Thanks.

**Comment by rogpeppe** [comment]
The NewDecoder definition is here:

https://github.com/go-yaml/yaml/blob/v2.2.1/yaml.go#L102

**Comment by niemeyer** [comment]
@dsoprea I've seen all your links and made very specific questions and suggestions that have not been replied to. As you say, it's a community and we're all just trying to help, but we cannot help if you completely ignore what we say.

**Comment by dsoprea** [comment]
@rogpeppe  Oh. Awesome. Thanks for pointing that out. I thought there might be a `NewDecoder`, but I figured it'd be somewhere adjacent to `newDecoder`, which was my failure for not grepping the tree. 

I see that you just added that a couple of months ago (https://github.com/go-yaml/yaml/commit/8e626dec39b5836cef636d885e33479debcf0cb1). I don't think I realized that I already had a copy of the project in my GOPATH and, obviously, I was on an older version.

Thanks a lot. I'll now return you to your normal programming....
